File: DSA/src/Questions/BackTracking/AdditiveNumbers.java
package Questions.BackTracking;
/*
https://leetcode.com/problems/additive-number/
 An additive number is a string whose digits can form an additive sequence.

A valid additive sequence should contain at least three numbers. Except for the first two numbers, each subsequent number in the sequence must be the sum of the preceding two.

Given a string containing only digits, return true if it is an additive number or false otherwise.

Note: Numbers in the additive sequence cannot have leading zeros, so sequence 1, 2, 03 or 1, 02, 3 is invalid.

 

Example 1:

Input: "112358"
Output: true
Explanation: 
The digits can form an additive sequence: 1, 1, 2, 3, 5, 8. 
1 + 1 = 2, 1 + 2 = 3, 2 + 3 = 5, 3 + 5 = 8
Example 2:

Input: "199100199"
Output: true
Explanation:
The additive sequence is: 1, 99, 100, 199.
1 + 99 = 100, 99 + 100 = 199


Constraints:

1 <= num.length <= 35
num consists only of digits.


Follow up: How would you handle overflow for very large input integers?
Time Complexity: O(n^3)
Space Complexity: O(1)


Base Case for Recursion
if (cur >= num.length()): If the current index is beyond the length of the string, check if at least three numbers have been used.
If so, return true. This ensures that the sequence has at least three numbers.

if (num.charAt(cur) == '0' && number1 + number2 != 0):
If the current character is '0' and the sum of the last two numbers is not zero, return false to avoid invalid sequences with leading zeros.

Loop for Generating Numbers
The loop inside the solve function generates the next number in the sequence by iterating through the remaining part of the string.
for (int i = cur; i < num.length(); i++):
number = number * 10 + num.charAt(i) - '0': Construct the next number by appending the current digit.
if (number == target && solve(number2, target, i + 1, num, count + 1)): If the constructed number matches the target (sum of the previous two numbers), recursively check the next part of the string.
else if (number > target): If the constructed number exceeds the target, break the loop as further digits will only increase the number.
 */
public class AdditiveNumbers {
    public boolean isAdditiveNumber(String num) {
        long number1 = 0;
        for (int i = 0; i < num.length() - 1; i++) {
            number1 = number1 * 10 + num.charAt(i) - '0';
            long number2 = 0;
            for (int j = i + 1; j < num.length(); j++) {
                number2 = number2 * 10 + num.charAt(j) - '0';
                if (solve(number1, number2, j + 1, num, 2))
                    return true;
                if (number2 == 0)
                    break;
            }
            if (number1 == 0)
                break;
        }
        return false;
    }

    boolean solve(long number1, long number2, int cur, String num, int count) {
        if (cur >= num.length()) {
            if (count >= 3)
                return true;
            return false;
        }
        if (num.charAt(cur) == '0' && number1 + number2 != 0)
            return false;

        long number = 0;
        long target = number1 + number2;
        for (int i = cur; i < num.length(); i++) {
            number = number * 10 + num.charAt(i) - '0';
            if (number == target && solve(number2, target, i + 1, num, count + 1))
                return true;
            else if (number > target)
                break;
        }
        return false;
    }
    public static void main(String[] args) {
        AdditiveNumbers additiveNumbers = new AdditiveNumbers();
        System.out.println(additiveNumbers.isAdditiveNumber("112358"));
        System.out.println(additiveNumbers.isAdditiveNumber("199100199"));
    }
}

File: DSA/Python/Questions/BackTracking/AdditiveNumbers.py
"""
# Question: Additive Number
# Link: https://leetcode.com/problems/additive-number/

# Check if string can form additive sequence

# Time Complexity: O(n^3)
# Space Complexity: O(1)

# Algorithm:
# 1. Generate first two numbers
# 2. Recursively validate sequence
# 3. Handle leading zeros
# 4. Check if sequence is valid

# Key Components:
# - is_additive_number(): Main validation implementation
# - solve(): Recursive helper for sequence validation
# - Number generation and validation logic
"""


class AdditiveNumbers:
    def is_additive_number(self, num: str) -> bool:
        number1 = 0
        for i in range(len(num) - 1):
            number1 = number1 * 10 + int(num[i])
            number2 = 0
            for j in range(i + 1, len(num)):
                number2 = number2 * 10 + int(num[j])
                if self.solve(number1, number2, j + 1, num, 2):
                    return True
                if number2 == 0:
                    break
            if number1 == 0:
                break
        return False

    def solve(self, number1: int, number2: int, cur: int, num: str, count: int) -> bool:
        if cur >= len(num):
            return count >= 3

        if num[cur] == "0" and number1 + number2 != 0:
            return False

        number = 0
        target = number1 + number2

        for i in range(cur, len(num)):
            number = number * 10 + int(num[i])
            if number == target and self.solve(number2, target, i + 1, num, count + 1):
                return True
            elif number > target:
                break

        return False


def main():
    solution = AdditiveNumbers()
    print(solution.is_additive_number("112358"))  # True
    print(solution.is_additive_number("199100199"))  # True


if __name__ == "__main__":
    main()
